allocation: create separate limit for frozen tier concurrent rebalance

schase-es opened the pull request:
The concurrent rebalance limit prevents too many shard relocations from happening at once. This works well in limiting active shards on the hot tier, so compute for indexing is reserved. Given that the frozen tier is an archive, compute separation is less of a concern and more rebalancing activity can occur without degradation. This change creates a separate variable for managing this.

Fixes: ES-11303

schase-es committed:
allocation: create separate limit for frozen tier concurrent rebalance

The concurrent rebalance limit prevents too many shard relocations from
happening at once. This works well in limiting active shards on the hot tier, so
compute for indexing is reserved. Given that the frozen tier is an archive,
compute separation is less of a concern and more rebalancing activity can occur
without degradation. This change creates a separate variable for managing this.

Commit:
[CI] Update transport version definitions

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es committed:
Formatting fixes

schase-es committed:
Adding unlimited test

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java:
```diff
@@ -68,18 +95,25 @@ public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation alloca
      */
     @Override
     public Decision canRebalance(RoutingAllocation allocation) {
```
As discussed, I think we want this to be 

```java
Decision canRebalance(allocation) {
   return frozenRebabalanceLimitExceeded(allocation) && regularRebalanceExceeded(allocation) ? THROTTLE : YES
}
```

nicktindall replied:
Could also just return `YES` here and throttle in the shard-specific method, but checking limits here gives an opportunity for early exit.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java:
```diff
@@ -89,12 +123,35 @@ public Decision canRebalance(RoutingAllocation allocation) {
+            }
+        }
+        return frozenRelocations;
+    }
```
`canRebalance(shard, allocation)` can be called once for a very large number of shards, so it might be worth maintaining this count in `RoutingNodes` a-la `relocatingShardCount` instead of recalculating it every time

nicktindall commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/ConcurrentRebalanceRoutingTests.java:
```diff
@@ -32,18 +35,41 @@
-                .put("cluster.routing.allocation.cluster_concurrent_rebalance", 3)
-                .build()
-        );
+        boolean testFrozen = randomBoolean();
```
it's probably worth having a separate method to test frozen? see https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#use-randomized-testing-for-coverage

nicktindall reviewed (commented):
Looking good, I think the approach is sound, using the different limits depending on the type of node.

schase-es committed:
Addressed or vs. and logic in the context of the allocation algorithm
- per-shard decision has been implemented, using the cluster-wide logic from before
- cluster-wide decision has been changed to allow progress whenever there is any budget

schase-es commented:
Thanks Nick -- this was really helpful.

I updated the overall algorithm shape to match what we talked about, where the shard-level one implements the cluster-wide logic from before, but specific to frozen or non-frozen counts. The cluster-level algorithm checks for room in either limit, including one bit we didn't mention but I realize mattered: one limit having "unlimited" as a configuration status.

I have some more to do on optimizing the counts for frozen. I need to have a look at this and decide what to do, but I think I need to keep a shardId set in routing nodes, or set a flag on ShardRouting.

There are a few other comments you made that I didn't get to quite yet.

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
- Fixed up cluster-wide algorithm for applying frozen tier limits
- Implemented count of relocatingShards inside RoutingNodes

schase-es committed:
Adding separate test for frozen

schase-es commented:
Hi Nick -- I added the count inside RoutingNodes, as discussed. It wasn't as bad as I thought.

I am wondering what you think about the cluster-wide logic. The check for unlimited makes things a little different than the addition we talked about, but I think what I have is simple enough.

It could do with a Decision refactor I guess.

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Check RoutingNode.node() for null in frozenNode test

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java:
```diff
@@ -68,33 +138,58 @@ public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation alloca
+            return true;
+        }
+        return false;
+    }
```
I would like to check with someone else, too.

I wrote the frozen/non-frozen counts, so that frozen is a subset count of all relocations. If a relocation isn't counted as frozen, then it's counted as non-frozen. So if something is mistaken, it's just subjected to a tighter constraint.

I can add a flag to ShardRouting, but as we discussed (and you're reminding me here) it may be too far for the data tier concept to propagate.

schase-es commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java:
```diff
@@ -44,21 +46,89 @@ public class ConcurrentRebalanceAllocationDecider extends AllocationDecider {
     @Override
     public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
-        return canRebalance(allocation);
+        int relocatingFrozenShards = allocation.routingNodes().getRelocatingFrozenShardCount();
```
Yes -- the non-frozen relocation count is the non-frozen count minus the frozen count. It's a little weird; I could separate them into two bins, rather than the frozen count being smaller. I was being careful to not modify the relocating count.

nicktindall replied:
Ah of course, you need them in both branches.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java:
```diff
@@ -245,6 +245,15 @@ public Stream<ShardRouting> shardsWithState(ShardRoutingState state) {
+     */
+    public int shardCountsWithState(ShardRoutingState state) {
+        return internalGetShardsWithState(state).size();
+    }
```
Is this the same as `numberOfShardsWithState` ?

schase-es replied:
Ah yes -- unsure how I missed that method.

schase-es committed:
Addressed review/extra feedback
- removed extraneous count method in RoutingNode
- renamed RoutingNodes.isFrozenNode -> isDedicatedFrozenNode
- centralized frozen node check: decider had own
- reuse original "reached the limit" log message langauge; this got mangled

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es committed:
Adding in mixed frozen and hot test

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

nicktindall reviewed (commented):
Looking good, mostly cosmetic comments

schase-es committed:
Test fixes

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es committed:
Set default concurrent rebalance for frozen to 2

schase-es committed:
Default to normal setting

schase-es committed:
Renaming strategy to allocationService

schase-es committed:
Fixed up decision message formatting, and added test

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es committed:
Slight improvements to the test structure

nicktindall reviewed (approved):
This LGTM with some nits

schase-es committed:
Docs fixes

schase-es committed:
Adding randomization

schase-es committed:
Refactoring state checks and settings

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es committed:
Formatting fixes

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java:
```diff
@@ -68,18 +95,25 @@ public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation alloca
             // Above allocator is used in DesiredBalanceComputer. Since we do not move actual shard data during calculation
             // it is possible to artificially set above setting to 2 to avoid unnecessary moves in desired balance.
+            // Separately: keep overall limit in simulation to two including frozen shards
             return allocation.decision(Decision.THROTTLE, NAME, "allocation should move one shard at the time when simulating");
```
@ywangd has begun the process of phasing this branch out.

https://github.com/elastic/elasticsearch/pull/136794/files#r2443798513

schase-es committed:
Merge branch 'main' into ES-11303_concurrent-frozen-tier-activity_rebase

schase-es committed:
Adding Locale.ROOT to all String.format calls

schase-es committed:
Undo Locale commit

schase-es committed:
Use Strings.format instead
